src/psyclone/tests/psyir/frontend/fparser2_where_handler_test.py

```diff
@@ -327,7 +327,11 @@ def test_where_within_loop(fortran_reader):
     assert isinstance(where_loop.loop_body[0], IfBlock)
     assign = where_loop.loop_body[0].if_body[0]
     assert isinstance(assign, Assignment)
-    assert [idx.name for idx in assign.lhs.indices] == ["widx1", "jl"]
+    assert assign.lhs.indices[0].debug_string() == "1 + widx1 - 1"
+    assert assign.lhs.indices[1].debug_string() == "jl"
+    assert where_loop.start_expr.value == "1"
+    assert (where_loop.stop_expr.debug_string() ==
+            "(UBOUND(var, dim=1) - LBOUND(var, dim=1)) / 1 + 1")
```

Comment:
Again, to improve the readability of the output, can we make the stop_expr `SIZE(var, dim=1)` when the lbound is known and it is "1"?

And I would like to see a separate test with a where statement with different lower bounds (known at compile time and not known), to see how the loop lower and upper bound are set then.

Also, with the mask dimension (which I think is used for the loop bounds) and the array in the body of the where with the range shifted (different lower bounds), as I think is one of the issues mentioned in the PR.

Reply:
Oh yes, now that I've done your suggestion and special-cased the "+1-1", I realise I don't have a test for anything else!

Reply:
I also realise that I've complicated things unnecessarily here:

    integer, dimension(4:5,-2:8) :: array

    write (*,*) 'First rank size = ', size(array, 1)
    write (*,*) 'Second rank size = ', size(array, 2)

gives:

    First rank size =            2
    Second rank size =           11

so I can use `SIZE` rather than `UBOUND` and `LBOUND`.

Reply:
In fact, it turned out that adding some tests revealed a quite significant bug - I'd been using the lower bound given by the `shape` of the datatype of the access and that is always either unknown or unity. (Because `shape` doesn't capture lower bound information.) I've fixed this now but also rediscovered #949 which means that we can't yet query a declaration if it has non-standard lower bounds. This means we have to resort to LBOUND more than we would like but that can be improved in future.